'odo describe' fails for component with two URLs with 'panic: runtime error: index out of range' error

/kind bug

## What versions of software are you using?

**Operating System:** 
Windows 10
**Output of `odo version`:**
odo v1.1.1 (02fbd3d08)

## How did you run odo exactly?
Not pushed git component with two URLs.

## Actual behavior
```
> odo.exe" describe
Component Name: nodejs-ex-git
Type: nodejs:10
Source: https://github.com/dgolovin/nodejs-ex
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/util.PrintComponentInfo(0xc0007afe00, 0xc0001cfab0, 0xd, 0x1befd6d, 0x9, 0x1c09821, 0x19, 0xc0001cfab0, 0xd, 0x0, ...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/util/cmdutils.go:144 +0x151a
github.com/openshift/odo/pkg/odo/cli/component.(*DescribeOptions).Run(0xc000830300, 0x0, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:86 +0x2f1
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x1f2a920, 0xc000830300, 0xc0008a8900, 0x3285110, 0x0, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:31 +0x143
github.com/openshift/odo/pkg/odo/cli/component.NewCmdDescribe.func1(0xc0008a8900, 0x3285110, 0x0, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:104 +0x65
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute(0xc0008a8900, 0x3285110, 0x0, 0x0, 0xc0008a8900, 0x3285110)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702 +0x290
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0006c9b00, 0x1d1bfd0, 0xc000074b40, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783 +0x2d1
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main()
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/cmd/odo/odo.go:75 +0x46a
```

## Expected behavior
Prints out two urls.

## Any logs, error output, etc?
N/A

--- a/pkg/component/component.go
+++ b/pkg/component/component.go
@@ -959,7 +959,10 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 		if len(urls) > 0 {
 			// We will clean up the existing value of ports and re-populate it so that we don't panic in `odo describe` and don't show inconsistent info
 			// This will also help in the case where there are more URLs created than the number of ports exposed by a component #2776
-			oldPortsProtocol := getPortsProtocolMapping(component.Spec.Ports)
+			oldPortsProtocol, err := getPortsProtocolMapping(component.Spec.Ports)
+			if err != nil {
+				return Component{}, err
+			}
 			component.Spec.Ports = []string{}
 
 			for _, url := range urls {
@@ -984,13 +987,17 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 // This function returns a mapping of port and protocol.
 // So for a value of ports {"8080/TCP", "45/UDP"} it will return a map {"8080":
 // "TCP", "45": "UDP"}
-func getPortsProtocolMapping(ports []string) map[string]string {
+func getPortsProtocolMapping(ports []string) (map[string]string, error) {
 	oldPortsProtocol := make(map[string]string, len(ports))
 	for _, port := range ports {
 		portProtocol := strings.Split(port, "/")
+		if len(portProtocol) != 2 {
+			// this will be the case if value of a port is something like 8080/TCP/something-else or simply 8080
+			return nil, errors.New("invalid <port/protocol> mapping. Please update the component configuration")
+		}
 		oldPortsProtocol[portProtocol[0]] = portProtocol[1]
 	}
-	return oldPortsProtocol
+	return oldPortsProtocol, nil
 }
 
 // ListIfPathGiven lists all available component in given path directory

--- a/pkg/component/component.go
+++ b/pkg/component/component.go
@@ -7,6 +7,7 @@ import (
 	"io"
 	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
 
 	"github.com/openshift/odo/pkg/devfile/adapters/common"
@@ -954,8 +955,18 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 			component.Spec.Source = util.GenFileURL(localConfig.GetSourceLocation())
 		}
 
-		for _, localURL := range localConfig.GetURL() {
-			component.Spec.URL = append(component.Spec.URL, localURL.Name)
+		urls := localConfig.GetURL()
+		if len(urls) > 0 {
+			// We will clean up the existing value of ports and re-populate it so that we don't panic in `odo describe` and don't show inconsistent info
+			// This will also help in the case where there are more URLs created than the number of ports exposed by a component #2776
+			oldPortsProtocol := getPortsProtocol(component.Spec.Ports)
+			component.Spec.Ports = []string{}
+
+			for _, url := range urls {
+				port := strconv.Itoa(url.Port)
+				component.Spec.Ports = append(component.Spec.Ports, fmt.Sprintf("%s/%s", port, oldPortsProtocol[port]))
+				component.Spec.URL = append(component.Spec.URL, url.Name)
+			}
 		}
 
 		for _, localEnv := range localConfig.GetEnvVars() {
@@ -968,6 +979,15 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 		return component, nil
 	}
 	return Component{}, nil
+}
+
+func getPortsProtocol(ports []string) map[string]string {
+	oldPortsProtocol := make(map[string]string, len(ports))
+	for _, port := range ports {
+		portProtocol := strings.Split(port, "/")
+		oldPortsProtocol[portProtocol[0]] = portProtocol[1]
+	}
+	return oldPortsProtocol
 }
 
 // ListIfPathGiven lists all available component in given path directory

--- a/pkg/component/component.go
+++ b/pkg/component/component.go
@@ -959,7 +959,7 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 		if len(urls) > 0 {
 			// We will clean up the existing value of ports and re-populate it so that we don't panic in `odo describe` and don't show inconsistent info
 			// This will also help in the case where there are more URLs created than the number of ports exposed by a component #2776
-			oldPortsProtocol := getPortsProtocol(component.Spec.Ports)
+			oldPortsProtocol := getPortsProtocolMapping(component.Spec.Ports)
 			component.Spec.Ports = []string{}
 
 			for _, url := range urls {
@@ -981,7 +981,10 @@ func GetComponentFromConfig(localConfig *config.LocalConfigInfo) (Component, err
 	return Component{}, nil
 }
 
-func getPortsProtocol(ports []string) map[string]string {
+// This function returns a mapping of port and protocol.
+// So for a value of ports {"8080/TCP", "45/UDP"} it will return a map {"8080":
+// "TCP", "45": "UDP"}
+func getPortsProtocolMapping(ports []string) map[string]string {
 	oldPortsProtocol := make(map[string]string, len(ports))
 	for _, port := range ports {
 		portProtocol := strings.Split(port, "/")

--- a/pkg/config/fakeConfig.go
+++ b/pkg/config/fakeConfig.go
@@ -18,9 +18,11 @@ func GetOneExistingConfigInfo(componentName, applicationName, projectName string
 	urlValue := []ConfigURL{
 		{
 			Name: "example-url-0",
+			Port: 8080,
 		},
 		{
 			Name: "example-url-1",
+			Port: 45,
 		},
 	}
 

--- a/pkg/config/fakeConfig.go
+++ b/pkg/config/fakeConfig.go
@@ -13,7 +13,7 @@ func GetOneExistingConfigInfo(componentName, applicationName, projectName string
 		},
 	}
 
-	portsValue := []string{"8080/TCP,45/UDP"}
+	portsValue := []string{"8080/TCP", "45/UDP"}
 
 	urlValue := []ConfigURL{
 		{

--- a/pkg/odo/cli/config/set.go
+++ b/pkg/odo/cli/config/set.go
@@ -33,6 +33,7 @@ var (
    %[1]s %[9]s 0.5
    %[1]s %[10]s 2
    %[1]s %[11]s 1
+   %[1]s %[12]s 8080/TCP,8443/TCP
 
    # Set a env variable in the local config
    %[1]s --env KAFKA_HOST=kafka --env KAFKA_PORT=6639
@@ -159,7 +160,7 @@ func NewCmdSet(name, fullName string) *cobra.Command {
 		Short: "Set a value in odo config file",
 		Long:  fmt.Sprintf(setLongDesc, config.FormatLocallySupportedParameters()),
 		Example: fmt.Sprintf(fmt.Sprint("\n", setExample), fullName, config.Type,
-			config.Name, config.MinMemory, config.MaxMemory, config.Memory, config.DebugPort, config.Ignore, config.MinCPU, config.MaxCPU, config.CPU),
+			config.Name, config.MinMemory, config.MaxMemory, config.Memory, config.DebugPort, config.Ignore, config.MinCPU, config.MaxCPU, config.CPU, config.Ports),
 		Args: func(cmd *cobra.Command, args []string) error {
 			if o.envArray != nil {
 				// no args are needed

--- a/tests/integration/component.go
+++ b/tests/integration/component.go
@@ -235,6 +235,7 @@ func componentTests(args ...string) {
 		It("should describe the component when it is not pushed", func() {
 			helper.CmdShouldPass("odo", append(args, "create", "nodejs", "cmp-git", "--project", project, "--git", "https://github.com/openshift/nodejs-ex", "--context", context, "--app", "testing")...)
 			helper.CmdShouldPass("odo", "url", "create", "url-1", "--context", context)
+			helper.CmdShouldPass("odo", "url", "create", "url-2", "--context", context)
 			helper.CmdShouldPass("odo", "storage", "create", "storage-1", "--size", "1Gi", "--path", "/data1", "--context", context)
 			helper.ValidateLocalCmpExist(context, "Type,nodejs", "Name,cmp-git", "Application,testing", "URL,0,Name,url-1")
 			cmpDescribe := helper.CmdShouldPass("odo", append(args, "describe", "--context", context)...)
@@ -242,12 +243,13 @@ func componentTests(args ...string) {
 			Expect(cmpDescribe).To(ContainSubstring("cmp-git"))
 			Expect(cmpDescribe).To(ContainSubstring("nodejs"))
 			Expect(cmpDescribe).To(ContainSubstring("url-1"))
+			Expect(cmpDescribe).To(ContainSubstring("url-2"))
 			Expect(cmpDescribe).To(ContainSubstring("https://github.com/openshift/nodejs-ex"))
 			Expect(cmpDescribe).To(ContainSubstring("storage-1"))
 
 			cmpDescribeJSON, err := helper.Unindented(helper.CmdShouldPass("odo", append(args, "describe", "-o", "json", "--context", context)...))
 			Expect(err).Should(BeNil())
-			expected, err := helper.Unindented(`{"kind": "Component","apiVersion": "odo.openshift.io/v1alpha1","metadata": {"name": "cmp-git","namespace": "` + project + `","creationTimestamp": null},"spec":{"app": "testing","type":"nodejs","source": "https://github.com/openshift/nodejs-ex","sourceType": "git","url": ["url-1"],"storage": ["storage-1"],"ports": ["8080/TCP"]},"status": {"state": "Not Pushed"}}`)
+			expected, err := helper.Unindented(`{"kind": "Component","apiVersion": "odo.openshift.io/v1alpha1","metadata": {"name": "cmp-git","namespace": "` + project + `","creationTimestamp": null},"spec":{"app": "testing","type":"nodejs","source": "https://github.com/openshift/nodejs-ex","sourceType": "git","url": ["url-1", "url-2"],"storage": ["storage-1"],"ports": ["8080/TCP"]},"status": {"state": "Not Pushed"}}`)
 			Expect(err).Should(BeNil())
 			Expect(cmpDescribeJSON).To(Equal(expected))
 			helper.CmdShouldPass("odo", append(args, "delete", "-f", "--all", "--context", context)...)

